catalogue/webapp/components/ImageEndpointSearchResults/ImageEndpointSearchResults.tsx

```diff
@@ -1,57 +1,126 @@
-import { FunctionComponent, useState } from 'react';
+import { FunctionComponent, useEffect, useState, useCallback } from 'react';
+import Gallery from 'react-photo-gallery';
+import styled from 'styled-components';
+
+import { Image, CatalogueResultsList } from '@weco/common/model/catalogue';
+
 import ExpandedImage from '../ExpandedImage/ExpandedImage';
 import ImageCard from '../ImageCard/ImageCard';
-import { Image, CatalogueResultsList } from '@weco/common/model/catalogue';
 import Modal from '@weco/common/views/components/Modal/Modal';
+import { convertImageUri } from '@weco/common/utils/convert-image-uri';
 
 type Props = {
   images: CatalogueResultsList<Image>;
 };
 
+type GalleryImageProps = Image & {
+  src: string;
+  width: number;
+  height: number;
+};
+
+const imageMargin = 16;
+
+const GalleryContainer = styled.div`
+  margin: 0 -${imageMargin / 2}px;
+
+  // This div is rendered by Gallery so we're targetting it on mobile only to ensure images are aligned on both sides
+  .react-photo-gallery--gallery > div {
+    justify-content: space-around;
+  }
+
+  @media (min-width: ${props => props.theme.sizes.medium}px) {
+    margin: 0 -${imageMargin}px;
+
+    .react-photo-gallery--gallery > div {
+      justify-content: flex-start;
+    }
+  }
+`;
+const ImageContainer = styled.li`
+  margin: 0 ${imageMargin ? imageMargin / 2 : 0}px
+    ${imageMargin ? imageMargin / 2 : 0}px;
+
+  @media (min-width: ${props => props.theme.sizes.medium}px) {
+    margin: 0 ${imageMargin}px ${imageMargin}px;
+  }
+`;
+
 const ImageEndpointSearchResults: FunctionComponent<Props> = ({
   images,
 }: Props) => {
   const [expandedImage, setExpandedImage] = useState<Image | undefined>();
+  const [data, setData] = useState<GalleryImageProps[]>([]);
+
   // In the case that the modal changes the expanded image to
   // be one that isn't on this results page, this index will be -1
   const expandedImagePosition = images.results.findIndex(
     img => expandedImage && img.id === expandedImage.id
   );
   const [isActive, setIsActive] = useState(false);
 
+  useEffect(() => {
+    if (data.length) setData([]);
+
+    images.results.map(async image => {
+      setData(prev => {
+        return [
+          ...prev,
+          {
+            ...image,
+            src: convertImageUri(image.locations[0].url, 300),
+            width: (image.aspectRatio || 1) * 100 + imageMargin,
+            height: 100,
+          },
+        ];
+      });
+    });
+  }, [images]);
```

Comment:
Not sure I understand the purpose of this: it seems like it does a _lot_ of state setting (ie rerendering), which React might be smart enough to batch but still feels like it might hurt performance. I might've missed something crucial in my reasoning (below), in which case I think we'd need some comments to explain rationale.

- `images` is a prop containing a list of Image entities from the catalogue API
- We want to add a `src`, `width`, and a `height` to all of these
- We are currently doing this by looping over the list, and for each Image we call `setData` to clone the old `data` list and append the new image to it. If we already have "old" `data` (I guess from the previous page), we first remove it.

Why do we need to use (a) the `data` state and (b) the `useEffect`? Could we just do the following?
```typescript
const imagesWithDimensions = images.results.map(image => ({
  ...image,
  src: convertImageUri(image.locations[0].url, 300),
  width: (image.aspectRatio || 1) * 100 + imageMargin,
  height: 100,
});
```
and then use `imagesWithDimensions` instead of `data`? I can see that this suggestion would recalculate whenever `setExpandedImage` was called, so we could further optimise it with [`useMemo`](https://reactjs.org/docs/hooks-reference.html#usememo):
```typescript
const imagesWithDimensions = useMemo(images.results.map(image => ({
  ...image,
  src: convertImageUri(image.locations[0].url, 300),
  width: (image.aspectRatio || 1) * 100 + imageMargin,
  height: 100,
}), [images.results]);
```

Reply:
A lot of great points; I'll try out some of these, think the `useMemo` makes a lot of sense straight off the bat.

> Will it definitely only affect the image search results page?

I've checked where the files I modified were used and it was the only place I could find (+ concept). Some of the components (like `IIIFImage`) are used elsewhere, but still supported in their former format.

> Are we happy with the bundle size increase?

What are the usual bounds for this? It was the one library I found that fit our needs so I'll admit that was the deciding factor.

Reply:
We don't have strict bounds, just common sense really in terms of it fitting our needs and not being outrageously large (if this was, say, 20kb, I might be more clearly opposed to it).

I think 6kb in this case is probably fine, my main concern would be making sure that it doesn't get pulled in on pages that don't use it (this is supposed to be automatic but sometimes common dependencies can make weird things happen). If you run a `yarn build` it should give you a nice little report of page bundle sizes + initial HTML sizes, does anything in there look suspicious (eg an increase on a page other than `/images`)?

Reply:
Alright just ran build on both this branch and main, only `images` and `concept` seem affected by it.

I've pushed more changes now having to do with how it looks on Concept without JS (transparent background looks better) and not to use the gallery if there are only one or two images (it looked wrong and it's not needed), like a concept can have.